File: node/runtime/src/cards.rs
use rstd::prelude::*;
use core::debug_assert;

pub type Nominal = u8;

pub const J: Nominal = 11;
pub const Q: Nominal = 12;
pub const K: Nominal = 13;
pub const A: Nominal = 1;

pub type Suit = u8;

pub const HEARTS:   Suit = 2;
pub const CLUBS:    Suit = 3;
pub const DIAMONDS: Suit = 4;
pub const SPADES:   Suit = 1;

#[derive(PartialEq, Eq, Debug)]
pub struct Card {
    pub nominal: Nominal,
    pub suit: Suit
}

impl Card {
    pub fn validate(&self) -> bool {
        debug_assert!(self.nominal >= 1 && self.nominal <= 13);
        debug_assert!(self.suit >= 1 && self.suit <= 4);

        self.nominal >= 1 && self.nominal <= 13
            && self.suit >= 1 && self.suit <= 4
    }
}

pub fn hearts(n: Nominal) -> Card {
    Card { nominal: n, suit: HEARTS }
}

pub fn clubs(n: Nominal) -> Card {
    Card { nominal: n, suit: CLUBS }
}

pub fn diamonds(n: Nominal) -> Card {
    Card { nominal: n, suit: DIAMONDS }
}

pub fn spades(n: Nominal) -> Card {
    Card { nominal: n, suit: SPADES }
}

pub fn encode(cards: &[Card]) -> Vec<u8> {
    cards.iter()
        .map(|card| {
            card.validate();
            vec![card.nominal, card.suit]
        })
        .flatten()
        .collect()
}

//a bit unfair random generation, see the test
pub fn from_random(bytes: &[u8]) -> Card {
    let card = Card {
        nominal: bytes[0] % 13 + 1,
        suit: bytes[1] % 4 + 1,
    };

    card.validate();
    card
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generate_from_random() {
        assert!(from_random(&[1, 0]) == spades(2));
        assert!(from_random(&[14, 0]) == spades(2));
        assert!(from_random(&[142, 6]) == clubs(K));

        assert!(from_random(&[0, 7]) == diamonds(A));
        assert!(from_random(&[247, 7]) == diamonds(A));
        assert!(from_random(&[248, 7]) == diamonds(2));
        assert!(from_random(&[255, 7]) == diamonds(9));
        //this means that 10, J, Q, K are less frequent

        assert!(from_random(&[255, 0]) == spades(9));
        assert!(from_random(&[255, 252]) == spades(9));
        assert!(from_random(&[255, 253]) == hearts(9));
        assert!(from_random(&[255, 254]) == clubs(9));
        assert!(from_random(&[255, 255]) == diamonds(9));
        //for suits, it is fair
    }
}

File: ui/src/cards.js
import React from 'react';

const J = 11;
const Q = 12;
const K = 13;
const A = 1;

const HEARTS   = 2;
const CLUBS    = 3;
const DIAMONDS = 4;
const SPADES   = 1;

export function decode (bytes) {
    let n = bytes.length;
    console.assert(n % 2 === 0,
        "Cards must be encoded with even number of bytes");

    return [...Array(n / 2).keys()]
        .map(i => { return {
            nominal: bytes[i * 2],
            suit: bytes[i * 2 + 1]
        }});
}

export function hidden () {
    return 'cards/back/bicycle_blue_mod.svg';
}

export function image (card) {
    return `cards/${suit(card.suit)}/${nominal(card.nominal)}.svg`;
}

function nominal (byte) {
    console.assert(byte <= 13,
        "Nominal cannot be encoded with number greater that 13");

    if (byte === J) { return "J"; }
    if (byte === Q) { return "Q"; }
    if (byte === K) { return "K"; }
    if (byte === A) { return "A"; }
    return byte.toString();
}

function suit (byte) {
    if (byte === HEARTS)   { return "hearts"; }
    if (byte === CLUBS)    { return "clubs"; }
    if (byte === DIAMONDS) { return "diamonds"; }
    if (byte === SPADES)   { return "spades"; }
}